safenode/src/protocol/types/transfers/mod.rs

```diff
@@ -0,0 +1,225 @@
+/// for the transactions to take effect.
+/// The peers will validate each signed spend they receive, before accepting it.
+/// Once enough peers have accepted all the spends of the transaction, and serve
+/// them upon request, the transaction will be completed.
+///
+/// (Disabled for now: Transfer fees will be added if not in data-network.)
+/// (Disabled for now: DbcReason, can be added later.)
+#[allow(clippy::result_large_err)]
+pub fn create_transfer(
+    dbcs: Vec<(Dbc, DerivedKey)>,
+    recipients: Vec<(Token, DbcIdSource)>,
+    change_to: PublicAddress,
+) -> Result<Outputs> {
+    // We need to select the necessary number of dbcs from those that we were passed.
+    // This will also account for any fees.
+    let send_inputs = select_inputs(dbcs, recipients, change_to)?;
+    transfer(send_inputs)
+}
+
+/// The tokens of the input dbcs will be transfered to the
+/// new dbcs (and a change dbc if any), which are returned from this function.
+/// This does not register the transaction in the network.
+/// To do that, the `signed_spends` of each new dbc, has to be uploaded
+/// to the network. When those same signed spends can be retrieved from
+/// enough peers in the network, the transaction will be completed.
+#[allow(clippy::result_large_err)]
+fn transfer(send_inputs: Inputs) -> Result<Outputs> {
+    let Inputs {
+        dbcs_to_spend,
+        recipients,
+        change: (change, change_to),
+    } = send_inputs;
+
+    let mut inputs = vec![];
+    for (dbc, derived_key) in dbcs_to_spend {
+        let revealed_amount = match dbc.revealed_amount(&derived_key) {
+            Ok(amount) => amount,
+            Err(err) => {
+                warn!("Ignoring dbc, as it didn't have the correct derived key: {err}");
+                continue;
+            }
+        };
+        let input = InputHistory {
+            input: RevealedInput::new(derived_key, revealed_amount),
+            input_src_tx: dbc.src_tx,
+        };
+        inputs.push(input);
+    }
+
+    let mut tx_builder = TransactionBuilder::default()
+        .add_inputs(inputs)
+        .add_outputs(recipients);
+
+    let mut rng = rng::thread_rng();
+
+    let dbc_id_src = change_to.random_dbc_id_src(&mut rng);
+    if change.as_nano() > 0 {
+        tx_builder = tx_builder.add_output(change, dbc_id_src);
+    }
+
+    // Finalize the tx builder to get the dbc builder.
+    let dbc_builder = tx_builder
+        .build(Hash::default(), &mut rng)
+        .map_err(Error::Dbcs)?;
+
+    // Perform verifications of input TX and spentproofs,
+    // as well as building the output DBCs.
+    let mut created_dbcs: Vec<_> = dbc_builder
+        .build()
+        .map_err(Error::Dbcs)?
+        .into_iter()
+        .map(|(dbc, amount)| CreatedDbc { dbc, amount })
+        .collect();
+
+    let mut change_dbc = None;
+    created_dbcs.retain(|created| {
+        if created.dbc.public_address() == &change_to && change.as_nano() > 0 {
```

Comment:
This is not 100% solid in my mind, since we could have another output with sending to ourselves.

In the wallet (which calls this code) such an operation is skipped.

But anyway..